### PR TITLE
support source view through @exposed_in decorator

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #97767
* #97766
* #97765
* __->__ #97764

Partially addresses https://github.com/pytorch/pytorch/issues/90284#issuecomment-1348941584. Source view was missing from functions decorated with `@exposed_in` since it relied on the `__module__` attribute to find the definition.

This PR implements a solution that fixes the source view without altering the object to be documented any further. An alternative solution could be to add an additional attribute to the decorated objects, e.g. `__source_module__` that stores the original path. If we want to go with this instead, the sphinx local extension implementation will be simpler, but as stated above, at the cost of increased non-standard surface.

cc @svekars @carljparker @zou3519 @Chillee @samdow @soumith @kshitij12345 @janeyx99

### CODE REVIEW

**Comment by pmeier** [comment]
[main](https://pytorch.org/docs/stable/generated/torch.func.jvp.html):

Screenshot from 2023-03-28 16-28-53

[PR](https://docs-preview.pytorch.org/97764/generated/torch.func.jvp.html):

Screenshot from 2023-03-28 16-28-20
587a-33a3-485d-b5a5-53a6aaa0969d.png)

**Review by albanD — APPROVED** [review]
Thanks for sharing the rendered doc links!

I'm not a fan of these source visualizer tbh, I'm not sure why they don't point to github links where it is much easier to navigate to other files @svekars do you know?

File: docs/source/conf.py [hunk 1]
```diff
@@ -579,6 +582,42 @@ def process_docstring(app, what_, name, obj, options, lines):
...
+    # relying on `__module__` and thus make the `@torch._functorch.utils.exposed_in` compatible with source view.
+
+    try:
+        obj = import_module(modname)
+        for part in fullname.split("."):
+            obj = getattr(obj, part)
+    except (ImportError, AttributeError):
+        # In case someone is using a partial build, some parts that should be documented are not accessible. `sphinx`
+        # warns in these cases, but otherwise continues the build. Thus, we do the same and just disable source view.
+        return None
```
Comment (zou3519) [inline docs/source/conf.py:598]: nit: should we raise a warning too?

File: docs/source/conf.py [hunk 2]
```diff
@@ -579,6 +582,42 @@ def process_docstring(app, what_, name, obj, options, lines):
...
+
+    fallback_module = getattr(obj, "__module__", None)
+
+    try:
+        abs_path = inspect.getsourcefile(obj)
+        if abs_path is None:
+            return None
+    except TypeError:
+        # `inspect.sourcefile` can only handle certain types. One noteable exception are "builtin methods or functions",
```
Comment (zou3519) [inline docs/source/conf.py:607]: nit: noteable -> notable

**Review by svekars — APPROVED** [review]

**Review by zou3519 — APPROVED** [review]
LGTM

**Comment by albanD** [comment]
@svekars ho yes for sure. My question was more why it points to custom page we generate and not directly at github?

**Pull request closed without merging**